src/engraving/dom/fret.cpp

```diff
@@ -127,48 +133,79 @@ EngravingItem* FretDiagram::linkedClone()
 std::shared_ptr<FretDiagram> FretDiagram::createFromString(Score* score, const String& s)
 {
     auto fd = Factory::makeFretDiagram(score->dummy()->segment());
-    int strings = static_cast<int>(s.size());
 
-    fd->setStrings(strings);
-    fd->setFrets(4);
-    fd->setPropertyFlags(Pid::FRET_STRINGS, PropertyFlags::UNSTYLED);
-    fd->setPropertyFlags(Pid::FRET_FRETS,   PropertyFlags::UNSTYLED);
-    int offset = 0;
-    int barreString = -1;
-    std::vector<std::pair<int, int> > dotsToAdd;
+    applyDiagramPattern(fd.get(), s);
 
-    for (int i = 0; i < strings; i++) {
-        Char c = s.at(i);
-        if (c == 'X' || c == 'O') {
-            FretMarkerType mt = (c == 'X' ? FretMarkerType::CROSS : FretMarkerType::CIRCLE);
-            fd->setMarker(i, mt);
-        } else if (c == '-' && barreString == -1) {
-            barreString = i;
-        } else {
-            int fret = c.digitValue();
-            if (fret != -1) {
-                dotsToAdd.push_back(std::make_pair(i, fret));
-                if (fret - 3 > 0 && offset < fret - 3) {
-                    offset = fret - 3;
+    return fd;
+}
+
+void FretDiagram::readHarmonyToDiagramFile(const muse::io::path_t& filePath)
+{
+    muse::io::File file(filePath);
+    if (!file.open()) {
+        LOGE() << file.errorString();
+        return;
+    }
+
+    XmlReader reader(&file);
```

Comment:
It might be worth adding a check that the map is empty:

```
IF_ASSERT_FAILED(s_harmonyToDiagramMap.empty()) {
    s_harmonyToDiagramMap.clear();
}
```